### PR TITLE
[PrintAsObjC] Print availability on classes, protocols, and categories

### PR DESCRIPTION
**Pull request by jrose-apple**
Now that Clang has availability diagnostics too (-Wpartial-availability and friends) we need to get this correct, or people will get warnings in the generated header!

Also fix up an issue with availability messages being omitted in certain cases.

rdar://problem/33313703

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test Linux

**Comment by jrose-apple** [comment]
I recommend reviewing this one commit-by-commit and/or with a diff that ignores whitespace: https://github.com/apple/swift/pull/11252/files?w=1

**Comment by jrose-apple** [comment]
```
error: argument labels '(extendingOrTruncating:)' do not match any available overloads
        return Int(extendingOrTruncating: result)
               ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

@moiseev, looks like you missed SwiftPM in your update?

**Comment by jrose-apple** [comment]
@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test Linux

**Comment by jrose-apple** [comment]
Review ping for @DougGregor or @devincoughlin.

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

File: lib/PrintAsObjC/PrintAsObjC.cpp [hunk 1]
```diff
@@ -710,132 +710,131 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
...
+      case PlatformKind::iOSApplicationExtension:
+        plat = "ios_app_extension";
+        break;
+      case PlatformKind::tvOSApplicationExtension:
+        plat = "tvos_app_extension";
+        break;
+      case PlatformKind::watchOSApplicationExtension:
+        plat = "watchos_app_extension";
+        break;
+      default:
```
Comment (DougGregor) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: Can we not have a `default` case here? The effect is the same, but we'll get warnings if `PlatformKind` ever gets extended.

**Review by DougGregor — APPROVED** [review]

File: lib/PrintAsObjC/PrintAsObjC.cpp (hunk 1 above)
Comment (jrose-apple) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: That's pre-existing code I just moved around, but sure.

Comment (DougGregor) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: Ah, ok.

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Pull request merged**